### PR TITLE
ETH-705 - send token2022 & new send service

### PR DESCRIPTION
**Pull request by eduardmaximovich**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/

## Description of Work

Sending token2022 and integrating new send service as replacedment for fee-relayer

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListView.kt [hunk 1]
```diff
@@ -89,38 +89,40 @@ class HistoryListView @JvmOverloads constructor(
...
-
-            is PagingState.Loading -> {
-                shimmerView.root.isVisible = isHistoryEmpty
-                refreshLayout.isVisible = true
-                errorStateLayout.root.isVisible = false
-                emptyStateLayout.root.isVisible = false
-                historyRecyclerView.isVisible = !isHistoryEmpty
-            }
+    override fun showPagingState(state: PagingState): Unit = with(binding) {
+        historyRecyclerView.post {
```
Comment (eduardmaximovich) [inline app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListView.kt:93]: fixed IllegalStateException

File: app/src/main/java/org/p2p/wallet/feerelayer/program/FeeRelayerProgram.kt [hunk 2]
```diff
@@ -258,6 +259,7 @@ object FeeRelayerProgram {
         return TransactionInstruction(feeRelayerProgramId, keys, bos.toByteArray())
     }
 
+    // todo: check program id !!
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/feerelayer/program/FeeRelayerProgram.kt:262]: ?

File: app/src/main/java/org/p2p/wallet/rpc/repository/account/RpcAccountRemoteRepository.kt [hunk 3]
```diff
@@ -71,10 +115,10 @@ class RpcAccountRemoteRepository(private val api: RpcAccountApi) : RpcAccountRep
 
         val programIds = listOf(
             TokenProgram.PROGRAM_ID,
-            TokenProgram.PROGRAM_ID_TOKEN2022
+            TokenProgram.TOKEN_2022_PROGRAM_ID
         )
 
-        val result = mutableListOf<TokenAccounts>()
+        val results = mutableListOf<Deferred<TokenAccounts>>()
```
Comment (eduardmaximovich) [inline app/src/main/java/org/p2p/wallet/rpc/repository/account/RpcAccountRemoteRepository.kt:121]: make request for SPL and Token2022 tokens

File: app/src/main/java/org/p2p/wallet/rpc/repository/account/RpcAccountRemoteRepository.kt [hunk 4]
```diff
@@ -71,10 +115,10 @@ class RpcAccountRemoteRepository(private val api: RpcAccountApi) : RpcAccountRep
 
         val programIds = listOf(
             TokenProgram.PROGRAM_ID,
-            TokenProgram.PROGRAM_ID_TOKEN2022
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/rpc/repository/account/RpcAccountRemoteRepository.kt:74]: i liked previous name more

File: app/src/main/java/org/p2p/wallet/rpc/repository/account/RpcAccountRemoteRepository.kt [hunk 5]
```diff
@@ -91,10 +135,14 @@ class RpcAccountRemoteRepository(private val api: RpcAccountApi) : RpcAccountRep
             )
 
             val rpcRequest = RpcRequest("getTokenAccountsByOwner", params)
-            result += api.getTokenAccountsByOwner(rpcRequest = rpcRequest).result
+            results += withContext(Dispatchers.IO) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/rpc/repository/account/RpcAccountRemoteRepository.kt:138]: why put withContext only here?

File: app/src/main/java/org/p2p/wallet/rpc/repository/solana/RpcSolanaRemoteRepository.kt [hunk 6]
```diff
@@ -28,6 +29,8 @@ class RpcSolanaRemoteRepository(
     private val gson: Gson
 ) : RpcSolanaRepository {
 
+    private var epochInfoCache: EpochInfo? = null
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/rpc/repository/solana/RpcSolanaRemoteRepository.kt:32]: check that RpcSolanaRemoteRepo is singleton

File: app/src/main/java/org/p2p/wallet/send/SendFeeRelayerManager.kt [hunk 7]
```diff
@@ -94,8 +117,21 @@ class SendFeeRelayerManager(
     fun buildTotalFee(
         sourceToken: Token.Active,
         calculationMode: CalculationMode,
+        tokenExtensions: TokenExtensionsMap,
     ): SendFeeTotal {
         val currentAmount = calculationMode.getCurrentAmount()
+
+        val transferFeePercent: BigDecimal? = tokenExtensions
+            .getTransferFeeConfig()
+            // todo: fix getting epoch, currently it doesn't work!
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/SendFeeRelayerManager.kt:126]: same todo

File: app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt [hunk 8]
```diff
@@ -165,7 +96,13 @@ class SendInteractor(
         val fees = userTokens
             .map { token ->
                 // converting SOL fee in token lamports to verify the balance coverage
-                async { getFeesInPayingTokenNullable(token, transactionFeeInSOL, accountCreationFeeInSOL) }
+                async {
+                    getFeesInPayingTokenUseCase.executeNullable(
+                        token,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt:101]: named args

File: app/src/main/java/org/p2p/wallet/send/SendFeeRelayerManager.kt [hunk 9]
```diff
@@ -94,8 +117,21 @@ class SendFeeRelayerManager(
...
+            ?.getActualTransferFee(networkObserver.getCurrentEpoch())
```
Comment (eduardmaximovich) [inline app/src/main/java/org/p2p/wallet/send/SendFeeRelayerManager.kt:127]: this should be replaced with SolanaRpcRepository.getEpochInfo().epoch, but it also needs to come up with how to call that suspended function or pass through current epoch using arguments instead

File: app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt [hunk 10]
```diff
@@ -271,59 +214,61 @@ class SendInteractor(
...
+    suspend fun sendTransactionV2(
+        destinationAddress: PublicKey,
+        token: Token.Active,
+        lamports: BigInteger
+    ): String = withContext(dispatchers.io) {
+        val sender = token.publicKey
 
-    suspend fun getUserRelayAccount(): RelayAccount =
-        feeRelayerAccountInteractor.getUserRelayAccount()
+        if (sender == destinationAddress.toBase58()) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt:224]: require(sender != destinationAddress.toBase58()) { .... }

File: app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt [hunk 11]
```diff
@@ -356,6 +301,20 @@ class SendInteractor(
                 recentBlockhash = recentBlockhash,
                 lamportsPerSignature = lamportsPerSignature
             )
+        } else if (token.isToken2022) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt:304]: when?

File: app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt [hunk 12]
```diff
@@ -523,4 +484,25 @@ class SendInteractor(
         val isSol = payingTokenMint == WRAPPED_SOL_MINT
         return noFreeTransactionsLeft && isSol
     }
+
+    private suspend fun decideWhoPaysNetworkFee(): Pair<SendFeePayerMode, Base58String?> {
+        // deciding what fees user will pay
+        val useFeeRelayer = !shouldUseNativeSwap(feePayerToken.mintAddress)
+
+        return if (useFeeRelayer) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt:492]: when?

File: app/src/main/java/org/p2p/wallet/send/interactor/usecase/GetTokenExtensionsUseCase.kt [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+import org.p2p.wallet.rpc.repository.account.RpcAccountRepository
+import org.p2p.wallet.send.repository.AccountInfoTokenExtensionsMapper.parseTokenExtensions
+
+class GetTokenExtensionsUseCase(
+    private val accountRepository: RpcAccountRepository,
+) {
+
+    suspend fun execute(token: Token.Active): TokenExtensionsMap {
+        val tokenInfo = accountRepository.getAccountInfoParsed(token.mintAddress)
+        return (tokenInfo?.parseTokenExtensions() ?: emptyMap()).also {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/interactor/usecase/GetTokenExtensionsUseCase.kt:15]: have you tried orEmpty here?

File: app/src/main/java/org/p2p/wallet/send/model/send_service/SendFeePayerMode.kt [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+import com.google.gson.annotations.SerializedName
+
+enum class SendFeePayerMode {
+    @SerializedName("Service")
+    Service,
+    @SerializedName("UserSol")
+    UserSol,
+    @SerializedName("UserSameToken")
+    UserSameToken,
+    Custom
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/model/send_service/SendFeePayerMode.kt:12]: @SerializedName like for NetworkFeeSource::User?

File: app/src/main/java/org/p2p/wallet/send/repository/AccountInfoTokenExtensionsMapper.kt [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+                    gson.fromJson(
+                        rawValue,
+                        AccountInfoTokenExtensionConfig.InterestBearingConfig::class.java
+                    )
+                }
+                else -> null
+            } ?: return@mapNotNull null
+
+            it.key to configResult
+        }.toMap()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/repository/AccountInfoTokenExtensionsMapper.kt:34]: have you tried associateBy with keySelector and ValueSelector?

File: app/src/main/java/org/p2p/wallet/send/repository/SendServiceRemoteRepository.kt [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            val rpcRequest = RpcRequest("get_compensation_tokens", params)
+            val response = api.getCompensationTokens(rpcRequest)
+            response.result
+        } catch (e: EmptyDataException) {
+            Timber.i("`get_compensation_tokens` responded with empty data, returning null")
+            emptyList()
+        }
+    }
+
+    override suspend fun generateTransactions(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/repository/SendServiceRemoteRepository.kt:31]:
```suggestion
    override suspend fun generateTransaction(
```

File: app/src/main/java/org/p2p/wallet/user/repository/UserTokensRemoteRepository.kt [hunk 17]
```diff
@@ -32,9 +32,10 @@ class UserTokensRemoteRepository(
     private suspend fun mapAccountsToTokens(publicKey: PublicKey, accounts: List<Account>): List<Token.Active> {
         val tokens = accounts.mapNotNull { token ->
             val mintAddress = token.account.data.parsed.info.mint
+            val programId = token.account.owner
 
             if (mintAddress == Constants.WRAPPED_SOL_MINT) {
-                // Hiding Wrapped Sol account because we are adding native SOL lower
+                // Hiding Wrapped Sol account because we are adding native SOL below
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/user/repository/UserTokensRemoteRepository.kt:38]: 😁

File: core/src/main/java/org/p2p/core/network/gson/GsonTypeAdapters.kt [hunk 18]
```diff
@@ -213,3 +217,43 @@ internal class OptionalTypeAdapter<T : Any>(
...
+        if (reader.peek() == JsonToken.NULL) {
+            reader.nextNull()
+            return null
+        }
+        val stringValue = reader.nextString()
+        return stringValue?.toBase64Instance()
+    }
+}
+
+internal class Base58StringTypeAdapter() : TypeAdapter<Base58String?>() {
```
Comment (gslevinkov) [inline core/src/main/java/org/p2p/core/network/gson/GsonTypeAdapters.kt:241]: have you checked org.p2p.core.network.gson.Base64TypeAdapter?

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/rpc/repository/account/RpcAccountRemoteRepository.kt (hunk 4 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/rpc/repository/account/RpcAccountRemoteRepository.kt:74]: Me too, but in official solana docs, they also call this TOKEN_2022_PROGRAM_ID, so I decided to standardize it, and of course we already have `ASSOCIATED_TOKEN_PROGRAM_ID` which also matches this naming: ${TYPE}_PROGRAM_ID

File: app/src/main/java/org/p2p/wallet/send/model/send_service/SendFeePayerMode.kt (hunk 14 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/send/model/send_service/SendFeePayerMode.kt:12]: Nope, it's internal enum which shouldn't be serialized or sent to the request

File: core/src/main/java/org/p2p/core/network/gson/GsonTypeAdapters.kt (hunk 18 above)
Comment (eduardmaximovich) [reply core/src/main/java/org/p2p/core/network/gson/GsonTypeAdapters.kt:241]: oh great! couldn't find it

File: app/src/main/java/org/p2p/wallet/feerelayer/program/FeeRelayerProgram.kt (hunk 2 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/feerelayer/program/FeeRelayerProgram.kt]: removed

File: app/src/main/java/org/p2p/wallet/rpc/repository/solana/RpcSolanaRemoteRepository.kt (hunk 6 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/rpc/repository/solana/RpcSolanaRemoteRepository.kt:32]: it's definitely singleton
```
single {
            val api = get<Retrofit>(named(RPC_RETROFIT_QUALIFIER)).create(RpcSolanaApi::class.java)
            RpcSolanaRemoteRepository(api, get(), get())
        } bind RpcSolanaRepository::class
```

File: app/src/main/java/org/p2p/wallet/rpc/repository/account/RpcAccountRemoteRepository.kt (hunk 5 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/rpc/repository/account/RpcAccountRemoteRepository.kt]: wrapped the entire function

File: app/src/main/java/org/p2p/wallet/user/repository/UserTokensRemoteRepository.kt (hunk 17 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/user/repository/UserTokensRemoteRepository.kt:38]: my hands were itching to fix that from the beginning, but had no reason to change this code

**Pull request merged**